kotlinx-coroutines-core/common/test/channels/ClosedChannelTest.kt

```diff
@@ -0,0 +1,196 @@
+        TestChannelKind.entries.forEach { kind ->
+            val channel = kind.create<Int>()
+            val iterator = channel.iterator()
+            channel.close()
+            assertFalse(iterator.hasNext())
+        }
+    }
+
+    @Test
+    fun testIteratorClosedWithExceptionHasNextThrows() = runTest {
+        TestChannelKind.entries.forEach { kind ->
+            val channel = kind.create<Int>()
+            val iterator = channel.iterator()
+            channel.close(TestException())
+            assertFailsWith<TestException> { (iterator.hasNext()) }
+        }
+    }
+
+    @Test
+    fun testClosedToListOk() = runTest {
+        TestChannelKind.entries.forEach { kind ->
+            val channel = kind.create<Int>()
+            channel.close()
+            channel.toList()
+        }
+    }
+
+    @Test
+    fun testClosedWithExceptionToListThrows() = runTest {
+        TestChannelKind.entries.forEach { kind ->
+            val channel = kind.create<Int>()
+            channel.close(TestException())
+            assertFailsWith<TestException> { channel.toList() }
+        }
+    }
+
+    @Test
+    fun testClosedConsumeEachOk() = runTest {
+        TestChannelKind.entries.forEach { kind ->
+            val channel = kind.create<Int>()
+            channel.close()
+            channel.consumeEach {  }
+        }
+    }
+
+    @Test
+    fun testClosedWithExceptionConsumeEachThrows() = runTest {
+        TestChannelKind.entries.forEach { kind ->
+            val channel = kind.create<Int>()
+            channel.close(TestException())
+            assertFailsWith<TestException> { channel.consumeEach {  } }
+        }
+    }
+
+    /**
+     * toList() alternative when a channel might close with an exception.
+     */
+
+    @Test
+    fun testToListCatching() = runTest {
+        TestChannelKind.entries.forEach { kind ->
+            val channel = kind.create<Unit>()
+            channel.close(TestException())
+            val lst = mutableListOf<Unit>()
+            while (true) {
+                channel.receiveCatching().onClosed { // suspends
+                    assertTrue(it is TestException)
+                    break
+                }.getOrThrow().apply(lst::add)
+            }
+            assertTrue(lst.isEmpty())
+        }
+    }
+
+    @Test
+    fun testTryToListCatching() = runTest {
+        TestChannelKind.entries.forEach { kind ->
+            val channel = kind.create<Unit>()
+            channel.close(TestException())
+            val lst = mutableListOf<Unit>()
+            while (true) {
+                channel.tryReceive().onClosed { // does not suspend
+                    assertTrue(it is TestException)
+                    break
+                }.onFailure {
+                    // empty, but not yet closed
+                    break
+                }.getOrThrow().apply(lst::add)
+            }
+            assertTrue(lst.isEmpty())
+        }
+    }
+
+    /**
```

Comment:
This syntax means the comment is a KDoc and refers to the entity directly below it. To make this a normal multiline comment for separating groups of tests, use `/*`.

That said, I think the separation is easy to break accidentally. Why not introduce separate classes to make this more robust?

```kotlin
class ClosedChannelTest: TestBase() {
    /**
     * Properties.
     *
     * Properties stay the same regardless of whether the channel was closed with or without exception.
     */
    class PropertiesTests: TestBase() {
        
    }
}
```